src/consumer/__tests__/consumeMessages.spec.js

```diff
@@ -205,6 +206,99 @@ describe('Consumer', () => {
     }
   })
 
+  it('after an error, aborts pending batch processing before retrying', async () => {
+    const partitionsConsumedConcurrently = 1
+    topicName = `test-topic-${secureRandom()}`
+    await createTopic({
+      topic: topicName,
+      partitions: partitionsConsumedConcurrently + 1,
+    })
+    await consumer.connect()
+    await producer.connect()
+    await consumer.subscribe({ topic: topicName, fromBeginning: true })
+
+    let inProgress = 0
+    let eachBatchCallCount = 0
+    let exceededConcurrencyLimit = false
+
+    const messagesConsumed = []
+    consumer.run({
+      partitionsConsumedConcurrently,
+      eachBatch: async ({ batch: { messages } }) => {
+        if (eachBatchCallCount++ === 0) throw new KafkaJSError(new Error('💣'), { retriable: true })
+        inProgress++
+        exceededConcurrencyLimit =
+          exceededConcurrencyLimit || inProgress > partitionsConsumedConcurrently
+        for (const event of messages) {
+          await sleep(2)
+          messagesConsumed.push(event)
+        }
+        inProgress--
+      },
+    })
+
+    await waitForConsumerToJoinGroup(consumer)
+
+    const messages = Array(100)
+      .fill()
+      .map(() => {
+        const value = secureRandom()
+        return { key: `key-${value}`, value: `value-${value}` }
+      })
+
+    await producer.send({ acks: 1, topic: topicName, messages })
+    await waitForMessages(messagesConsumed, { number: messages.length })
+    expect(messagesConsumed.length).toEqual(messages.length)
+    expect(exceededConcurrencyLimit).toBeFalse()
+  })
+
+  it('after an error, completes in-progress batch processing before retrying', async () => {
```

Comment:
This one is interesting. Given 3 partitions and a concurrency limit of 2. If partition 2 throws and we allow partition 1 to finish before throwing and retrying, does it mean that partition 1 gets re-processed, or do the offsets for partition 1 get resolved so that we skip those and only actually do any re-work for partition 2?

Again, this test needs some more explicit assertions to check that it's actually verifying what it says it's verifying. If the above is true, then there should be some way of asserting that partition 1 was only processed once.

Reply:
Re-worked the tests to be more focused and explicit